File: src/main/scala/crusader728/leetcode/tree/FindLeavesOfBinaryTree366.scala
package scala.crusader728.leetcode.tree

import scala.crusader728.leetcode.TreeNode

object FindLeavesOfBinaryTree366 {
  def findLeaves(root: TreeNode): List[List[Int]] = {
    val (h, depthMap) = collectHeight(root)
    (0 to h).foldRight(List.empty[List[Int]]) {case (i, acc) => {
      if(depthMap.contains(i)) {
        depthMap(i).map(_.value) :: acc
      } else {
        acc
      }
    }}
  }


  def collectHeight(root: TreeNode): (Int, Map[Int, List[TreeNode]]) = root match {
    case null => (-1, Map.empty)
    case n if n.left == null && n.right == null => (0, Map(0 -> List(n)))
    case n => {
      val (h1, m1): (Int, Map[Int, List[TreeNode]]) = if(n.left == null) {
        (-1, Map.empty)
      } else {
        collectHeight(n.left)
      }

      val (h2, m2): (Int, Map[Int, List[TreeNode]]) = if(n.right == null) {
        (-1, Map.empty)
      } else {
        collectHeight(n.right)
      }

      val h = Math.max(h1, h2) + 1
      val depthMap = m1.foldLeft(m2) {case (acc, (k, l)) => {
        acc + (k -> (acc.getOrElse(k, List.empty) ++ l))
      }}
      (h, depthMap + (h -> List(root)))
    }
  }

  def main(args: Array[String]): Unit = {
    println(findLeaves(new TreeNode(1,
      new TreeNode(2,
        new TreeNode(4, null, null),
        new TreeNode(5, null, null)
      ),
      new TreeNode(3, null, null)
    )))
  }
}

File: src/main/withJava/crusader728/leetcode/dfs/FindLeavesOfBinaryTree366.java
package withJava.crusader728.leetcode.dfs;

import java.util.ArrayList;
import java.util.List;

public class FindLeavesOfBinaryTree366 {
    public List<List<Integer>> findLeaves(TreeNode root) {
        if(root == null) {
            return new ArrayList<>();
        }

        List<List<Integer>> result = new ArrayList<>();
        dfs(root, result);
        return result;
    }

    private int dfs(TreeNode root, List<List<Integer>> result) {
        if(root.left == null && root.right == null) {
            if(result.isEmpty()) {
                result.add(new ArrayList<>());
            }
            result.get(0).add(root.val);
            return 0;
        } else {
            int left = -1;
            int right = -1;
            if(root.left != null) {
                left = dfs(root.left, result);
            }
            if(root.right != null) {
                right = dfs(root.right, result);
            }
            int height = Math.max(left, right) + 1;
            if(height >= result.size()) {
                result.add(new ArrayList<>());
            }
            result.get(height).add(root.val);
            return height;
        }
    }

    private class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
    }
}
